spec/app/ics-028-cross-chain-valset-update/README.md

```diff
@@ -0,0 +1,257 @@
+    lastObserved = content.last()
+
+    // return the entire content of content starting from oldLastObserved (excluding oldLastObserved)
+    return content.startFrom(oldLastObserved)
+  }
+}
+```
+
+- Expected precondition
+  - None
+- Expected postcondition
+  - Returns the elements enqueued since the last Q.read() operation was invoked.
+- Error condition
+  - None
+
+```golang
+// End-Block method executed at the end of each block
+func endBlock(block: Block) {
+  // get time
+  time = block.time
+
+  // finish unbonding for mature validator sets
+  while (!UnbondingQueue.isEmpty()) {
+    // peak the first queue entry
+    seqNum, startTime = UnbondingQueue.peak()
+
+    if (startTime + UNBONDING_PERIOD >= time) {
+      // remove from the unbonding queue
+      seqNum, startTime = UnbondingQueue.dequeue()
+
+      // dequeue all elements until seqNum
+      for (i = dequeueSeqNum + 1; i <= seqNum; i++) {
+        Q.dequeue()
+      }
+      dequeueSeqNum = seqNum
+
+    } else {
+      break
+    }
+  }
+
+  // get the new changes
+  changes = observeChangesQ()
+
+  // if there are no changes, return currnt validator set
+  if (changes.empty()) {
+    // validator set remains the same
+    ABCI.updateValidatorSet(block.validatorSet)
+    return
+  }
+
+  // get the old validator set
+  oldValSet = block.validatorSet
+
+  // get the new validator set; init to the old one
+  newValSet = oldValSet
+
+  // start unbonding for the old validator set represented by the validator set update;
+  // "start unbonding" simply means adding to the queue of validator set changes that started unbonding
+  UnbondingQueue.enqueue(validatorSetSeqNum, time)
+
+  // update the validator set
+  while (!changes.isEmpty()) {
+    valSetUpdate = changes.dequeue()
+
+    // update the new validator set
+    newValSet = applyValidatorUpdate(newValSet, valSetUpdate)
+
+    // increment validatorSetSeqNum
+    validatorSetSeqNum++
+
+    // remember which demands participate
+    if (content.isEmpty()) {
+      newSeqNum = validatorSetSeqNum
+    }
+  }
+
+  return newValSet
+}
+```
+
+- Expected precondition
+  - Every transaction from the *block* is executed
+- Expected postcondition
+  - Unbonding starts for the old validator set
+  - Unbonding finishes for all validator set that started unbonding more than *unbondingTime* before *time = block.time*. Moreover, the *UnbondingOver* packet is created for each such validator set
+  - The new validator set *newValSet* is pushed to the Tendermint protocol and *newValSet* reflects all the change validator set demands from the *block*
+- Error condition
+  - If the precondition is violated
+
+
+## Correctness Arguments
+
+Here we provide correctness arguments for the liveness, validator set change safety and liveness properties.
+
+### Liveness
+Suppose that the IBC communication indeed successfully relays the change validator set demand to the baby blockchain.
```

Comment:
There should be a synchrony bound here, no? What about timeouts?

Reply:
Yes, the synchrony bound is the IBC timeout.

Reply:
Can we clarify that?